File: Sources/MEMO/MemoStorage.cs
using System.Text;
using System.Text.Json;

namespace Memo;

public sealed class MemoStorage
{
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    private readonly string baseDir;
    private readonly string tabsDir;
    private readonly string manifestPath;
    private readonly string settingsPath;

    public MemoStorage()
    {
        string appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        baseDir = Path.Combine(appData, "MEMO");
        tabsDir = Path.Combine(baseDir, "tabs");
        manifestPath = Path.Combine(baseDir, "tabs.json");
        settingsPath = Path.Combine(baseDir, "settings.json");
    }

    public string BaseDirectory => baseDir;

    public MemoSession LoadSession()
    {
        try
        {
            if (File.Exists(manifestPath))
            {
                var session = JsonSerializer.Deserialize<MemoSession>(File.ReadAllText(manifestPath));
                if (session != null)
                {
                    var tabs = NormalizedTabs(session.Tabs);
                    if (tabs.Count > 0)
                    {
                        string? selectedId = tabs.Any(t => t.Id == session.SelectedTabId)
                            ? session.SelectedTabId
                            : tabs[0].Id;
                        return new MemoSession { Tabs = tabs, SelectedTabId = selectedId };
                    }
                }
            }
        }
        catch
        {
        }

        return new MemoSession { Tabs = new List<MemoTab> { MemoTab.Create(1) }, SelectedTabId = null };
    }

    public void SaveSession(MemoSession session)
    {
        Directory.CreateDirectory(baseDir);
        AtomicWrite(manifestPath, JsonSerializer.Serialize(session, JsonOptions));
    }

    public string? LoadTabRtf(string id) => TryRead(TabPath(id, ".rtf"));

    public string? LoadTabText(string id) => TryRead(TabPath(id, ".txt"));

    public void SaveTab(string id, string rtf, string plainText)
    {
        Directory.CreateDirectory(tabsDir);
        AtomicWrite(TabPath(id, ".rtf"), rtf);
        AtomicWrite(TabPath(id, ".txt"), plainText);
    }

    public void DeleteTab(string id)
    {
        try { File.Delete(TabPath(id, ".rtf")); } catch { }
        try { File.Delete(TabPath(id, ".txt")); } catch { }
    }

    public MemoSettings LoadSettings()
    {
        try
        {
            if (File.Exists(settingsPath))
            {
                var settings = JsonSerializer.Deserialize<MemoSettings>(File.ReadAllText(settingsPath));
                if (settings != null)
                {
                    return settings;
                }
            }
        }
        catch
        {
        }

        return new MemoSettings();
    }

    public void SaveSettings(MemoSettings settings)
    {
        Directory.CreateDirectory(baseDir);
        AtomicWrite(settingsPath, JsonSerializer.Serialize(settings, JsonOptions));
    }

    private static List<MemoTab> NormalizedTabs(List<MemoTab> tabs)
    {
        var seen = new HashSet<string>();
        var result = new List<MemoTab>();

        foreach (var tab in tabs)
        {
            string id = NormalizedId(tab.Id);
            if (id.Length == 0 || !seen.Add(id))
            {
                continue;
            }

            string title = (tab.Title ?? "").Trim();
            result.Add(new MemoTab
            {
                Id = id,
                Title = title.Length == 0 ? $"메모 {seen.Count}" : title,
                BodyIsSeparate = tab.BodyIsSeparate,
            });
        }

        return result;
    }

    private static string NormalizedId(string id)
    {
        var builder = new StringBuilder(id.Length);
        foreach (char c in id)
        {
            if (char.IsLetterOrDigit(c) || c == '-' || c == '_')
            {
                builder.Append(c);
            }
        }

        return builder.ToString();
    }

    private string TabPath(string id, string extension) =>
        Path.Combine(tabsDir, NormalizedId(id) + extension);

    private static string? TryRead(string path)
    {
        try
        {
            return File.Exists(path) ? File.ReadAllText(path) : null;
        }
        catch
        {
            return null;
        }
    }

    private static void AtomicWrite(string path, string content)
    {
        string tmp = path + ".tmp";
        File.WriteAllText(tmp, content, new UTF8Encoding(false));
        File.Move(tmp, path, overwrite: true);
    }
}

File: Sources/MEMO/MemoStorage.swift
import AppKit

final class MemoStorage {
    private let fileManager = FileManager.default
    private let directoryURL: URL
    private let tabsDirectoryURL: URL
    private let manifestURL: URL
    private let legacyRTFURL: URL
    private let legacyPlainTextURL: URL

    init() {
        let supportURL = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
        let baseURL = supportURL ?? fileManager.homeDirectoryForCurrentUser

        directoryURL = baseURL.appendingPathComponent("MEMO", isDirectory: true)
        tabsDirectoryURL = directoryURL.appendingPathComponent("tabs", isDirectory: true)
        manifestURL = directoryURL.appendingPathComponent("tabs.json", isDirectory: false)
        legacyRTFURL = directoryURL.appendingPathComponent("memo.rtf", isDirectory: false)
        legacyPlainTextURL = directoryURL.appendingPathComponent("memo.txt", isDirectory: false)
    }

    func loadSession() -> MemoSession {
        if let data = try? Data(contentsOf: manifestURL),
           let session = try? JSONDecoder().decode(MemoSession.self, from: data) {
            let tabs = normalizedTabs(session.tabs)

            if !tabs.isEmpty {
                let selectedID = tabs.contains(where: { $0.id == session.selectedTabID }) ? session.selectedTabID : tabs.first?.id
                return MemoSession(tabs: tabs, selectedTabID: selectedID)
            }
        }

        return MemoSession(tabs: [MemoTab.create(index: 1)], selectedTabID: nil)
    }

    func saveSession(_ session: MemoSession) throws {
        try fileManager.createDirectory(at: directoryURL, withIntermediateDirectories: true)

        let data = try JSONEncoder().encode(session)
        try data.write(to: manifestURL, options: [.atomic])
    }

    func load(tab: MemoTab) -> NSAttributedString? {
        let rtfURL = tabRTFURL(for: tab.id)
        let plainTextURL = tabPlainTextURL(for: tab.id)

        if let rtfData = try? Data(contentsOf: rtfURL),
           let attributedText = NSAttributedString(rtf: rtfData, documentAttributes: nil) {
            return attributedText
        }

        guard let plainText = try? String(contentsOf: plainTextURL, encoding: .utf8) else {
            return loadLegacyMemoIfNeeded()
        }

        return NSAttributedString(string: plainText)
    }

    func save(_ attributedText: NSAttributedString, for tab: MemoTab) throws {
        try fileManager.createDirectory(at: tabsDirectoryURL, withIntermediateDirectories: true)

        let fullRange = NSRange(location: 0, length: attributedText.length)
        let rtfData = try attributedText.data(
            from: fullRange,
            documentAttributes: [.documentType: NSAttributedString.DocumentType.rtf]
        )

        try rtfData.write(to: tabRTFURL(for: tab.id), options: [.atomic])
        try attributedText.string.write(to: tabPlainTextURL(for: tab.id), atomically: true, encoding: .utf8)
    }

    func delete(tab: MemoTab) {
        try? fileManager.removeItem(at: tabRTFURL(for: tab.id))
        try? fileManager.removeItem(at: tabPlainTextURL(for: tab.id))
    }

    private func loadLegacyMemoIfNeeded() -> NSAttributedString? {
        if hasStoredTabFiles() {
            return nil
        }

        if let rtfData = try? Data(contentsOf: legacyRTFURL),
           let attributedText = NSAttributedString(rtf: rtfData, documentAttributes: nil) {
            return attributedText
        }

        guard let plainText = try? String(contentsOf: legacyPlainTextURL, encoding: .utf8) else {
            return nil
        }

        return NSAttributedString(string: plainText)
    }

    private func hasStoredTabFiles() -> Bool {
        guard let files = try? fileManager.contentsOfDirectory(at: tabsDirectoryURL, includingPropertiesForKeys: nil) else {
            return false
        }

        return files.contains { file in
            let pathExtension = file.pathExtension.lowercased()
            return pathExtension == "rtf" || pathExtension == "txt"
        }
    }

    private func normalizedTabs(_ tabs: [MemoTab]) -> [MemoTab] {
        var seen = Set<String>()

        return tabs.compactMap { tab in
            let id = normalizedID(tab.id)
            guard !id.isEmpty, !seen.contains(id) else {
                return nil
            }

            seen.insert(id)
            let title = tab.title.trimmingCharacters(in: .whitespacesAndNewlines)
            return MemoTab(
                id: id,
                title: title.isEmpty ? "메모 \(seen.count)" : title,
                hasSeparateTitle: tab.hasSeparateTitle
            )
        }
    }

    private func normalizedID(_ id: String) -> String {
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-_"))
        let scalars = id.unicodeScalars.filter { allowed.contains($0) }
        return String(String.UnicodeScalarView(scalars))
    }

    private func tabRTFURL(for id: String) -> URL {
        tabsDirectoryURL.appendingPathComponent("\(normalizedID(id)).rtf", isDirectory: false)
    }

    private func tabPlainTextURL(for id: String) -> URL {
        tabsDirectoryURL.appendingPathComponent("\(normalizedID(id)).txt", isDirectory: false)
    }
}
